protected_app_signals_cost.md

```diff
@@ -0,0 +1,168 @@
+
+The distribution of the cost for running the Bidding Server based on internal estimates is shown below:
+
+The three factors are:
+
+*   [Inference][15]. This is expected to constiture the vast majority of the cost. Details on how to compute the inference costs will be added later.
+*   Cost of preparing the data for ad retrieval: essentially the cost for running the. prepareDataForAdRetrievalUDF. This is a minor contributor.
+*   Base cost for running the bidding process, this is the cost for coordinating the flow, processing the data from the ads retrieval server and sending it to generateBid, handling the response and returning the results to the BFE.
+
+There are a lot of changes in bidding server
+
+1.  prepareDataForAdRetrieval - This is a new function that runs before generateBid when using TEE based retrieval and it is skipped in case of contextual based retrieval. The function is intended to be used to decode the signals received from the device and make them available to the retrieval server UDF and to the generateBid. We are expecting that this function might run some inference too. The main factor to consider when estimating this cost are:
+    1.  Complexity of the model used for feature extraction
+    2.  Amount of signals received (will likely always be the currently max allowed size of 1.5KB but we are working on enabling the seller to control the payload size alloted for each buyer so this might change).
+    3.  Logic and language used to implement the UDF (using JS or WASM)
+    4.  Cost of communicating with the inference service (sidecar process at the moment)
+2.  Sending request to retrieval server (marked as O in the diagram above) - cost is zero
+3.  Getting info back from retrieval and sending it to generateBid. This may involve data copying into v8. We need to figure out how much data this is, and what the copying costs are (response size from retrieval, size of embeddings, number of embeddings, number of ads).
+4.  generateBid:
+    1.  Cost of the script itself - maybe assume this is similar to the remarketing case
+    2.  Cost of sending inference requests to inference sidecar + cost per inference. To compute this you need to know how many ads will be batched in a single inference, how many models will be used during bidding and preparation of data for retrieval, the, size of each model and how expensive each inference call is.
+    3.  Compared to PA, where each Custom Audience will run generateBid with its subset of ads, there is only one generateBid invocation for all the PAS scripts
+5.  General inference cost - download models, general bookkeeping (memory cost of storing models in sidecars). Model storage cost on gcs.
```

Comment:
GCS gives an impression that this explainer is applicable to GCP only but rest of the explainer seems generic.

Reply:
very good point. this was a legacy of the first design